src/rulesets/SparseArrays/sparsematrix.jl

```diff
@@ -10,16 +10,52 @@ function rrule(::typeof(sparse), I::AbstractVector, J::AbstractVector, V::Abstra
     return sparse(I, J, V, m, n, combine), sparse_pullback
 end
 
-function rrule(::Type{T}, A::AbstractMatrix) where T <: SparseMatrixCSC
+function rrule(::Type{T}, A::AbstractMatrix) where T <: AbstractSparseMatrix
     function sparse_pullback(Ω̄)
         return NoTangent(), Ω̄
     end
     return T(A), sparse_pullback
 end
 
-function rrule(::Type{T}, v::AbstractVector) where T <: SparseVector
+function rrule(::Type{T}, v::AbstractVector) where T <: AbstractSparseVector
     function sparse_pullback(Ω̄)
         return NoTangent(), Ω̄
     end
     return T(v), sparse_pullback
 end
+
+function rrule(::typeof(findnz), A::AbstractSparseMatrix)
+    I, J, V = findnz(A)
+    m, n = size(A)
+
+    function findnz_pullback(Δ)
+        _, _, V̄ = unthunk(Δ)
+        V̄ isa AbstractZero && return (NoTangent(), V̄)
+        return NoTangent(), sparse(I, J, V̄, m, n)
+    end
+
+    return (I, J, V), findnz_pullback
+end
+
+function rrule(::typeof(findnz), v::AbstractSparseVector)
+    I, V = findnz(v)
+    n = length(v)
+
+    function findnz_pullback(Δ)
+        _, V̄ = unthunk(Δ)
+        V̄ isa AbstractZero && return (NoTangent(), V̄)
+        return NoTangent(), sparsevec(I, V̄, n)
+    end
+
+    return (I, V), findnz_pullback
+end
+
+function rrule(::typeof(Broadcast.broadcasted), T::Type{<:Number}, x::AbstractSparseArray)
+    proj = ProjectTo(x)
+    function broadcasted_cast_sparse(Δ)
+        return NoTangent(), NoTangent(), proj(Δ)         
+    end
+    return T.(x), broadcasted_cast_sparse
+end
```

Comment:
I think this is the first broadcast rule in here. I'm not entirely sure how I feel about this. 

`T.(x)` is eager, would there be downsides to leaving it lazy, like `broadcasted(T, x)`? That may mean more things fuse, or it may mean that subsequent Zygote broadcasting rules fail because they want an array?

Similarly `proj(Δ)` won't accept something like a broadcasted, if we explore fused gradients that way. 

These are the sort of questions that make me unsure we want to commit to one design here. I guess having one or two won't hurt too much, we can fix them here later if needed.

Zygote's present rule fails, I believe it's trying to make a sparsearray of tuples? 
```julia
julia> gradient(x -> sum(Float32.(x)), sprand(10, 0.5))
ERROR: MethodError: no method matching zero(::Type{Any})
Closest candidates are:
  zero(::Type{Union{Missing, T}}) where T at ~/.julia/dev/julia/usr/share/julia/base/missing.jl:105
  zero(::Union{Type{P}, P}) where P<:Dates.Period at ~/.julia/dev/julia/usr/share/julia/stdlib/v1.8/Dates/src/periods.jl:53
  zero(::CartesianIndex{N}) where N at ~/.julia/dev/julia/usr/share/julia/base/multidimensional.jl:106
  ...
Stacktrace:
  [1] zero(#unused#::Type{Any})
    @ Base ./missing.jl:106
  [2] _zeros_eltypes
    @ ~/.julia/dev/julia/usr/share/julia/stdlib/v1.8/SparseArrays/src/higherorderfns.jl:208 [inlined]
  [3] _noshapecheck_map(::typeof(first), ::SparseVector{Any, Int64})
    @ SparseArrays.HigherOrderFns ~/.julia/dev/julia/usr/share/julia/stdlib/v1.8/SparseArrays/src/higherorderfns.jl:164
  [4] map(f::typeof(first), A::SparseVector{Any, Int64})
    @ SparseArrays.HigherOrderFns ~/.julia/dev/julia/usr/share/julia/stdlib/v1.8/SparseArrays/src/higherorderfns.jl:147
  [5] adjoint
    @ ~/.julia/packages/Zygote/FPUm3/src/lib/broadcast.jl:187 [inlined]
  [6] _pullback(__context__::Zygote.Context, 714::typeof(Base.Broadcast.broadcasted), 715::SparseArrays.HigherOrderFns.SparseVecStyle, f::Type{Float32}, args::SparseVector{Float64, Int64})
    @ Zygote ~/.julia/packages/ZygoteRules/AIbCs/src/adjoint.jl:65
```
But perhaps Zygote ought to have a rule for broadcasting `(::Type).(x)`, since what's written here would work for arbitrary arrays I believe, not just sparse, and would be faster than what it does now.

Reply:
maybe it's ok to have the first broadcast here covering a limited use case not currently supported by Zygote and see how it goes? There are plenty of issues related to sparse arrays in the ML ecosystem, I'd like to get things going for downstream packages

Reply:
Would merging the `findnz` and separating broadcasting out in a separate PR be an option? Since `findnz` is complete and broadcasting possibly requires in-depth discussion?

Reply:
ok